### PR TITLE
ENH: improve precision of genlogistic methods

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Related to #17832

#### What does this implement/fix?
Overrides for `genlogistic`:
* `logcdf`
* `sf=-expm1(logcdf)`
* `isf`

Improves precision of `ppf` near 1 using the new `powm1` function from special.

#### Additional information
*PPF*

image

image

<details>

```
import numpy as np
from scipy import stats
from scipy import special as sc
import matplotlib.pyplot as plt

mp.dps = 500

@np.vectorize
def genlogistic_ppf_mpmath(q, c):
    q = mp.mpf(q)
    c = mp.mpf(c)
    return float(-mp.log(q**(-mp.one/c) - mp.one))

def genlogistic_ppf(q, c):
    return -np.log(sc.powm1(q, -1.0/c))

def genlogistic_isf_pr(x, c):
    return -np.log(sc.expm1(-np.log(-sc.expm1(np.log(x)))/c))

x = np.linspace(1e-30, 0.9999999999, 1000)
c = 0.2
reference = np.asarray(genlogistic_ppf_mpmath(x, c), np.float64)
main_result = stats.genlogistic.ppf(x, c)
pr_result = genlogistic_ppf(x, c)

plt.semilogy(x, np.abs((main_result - reference)/reference), label="main", alpha=0.5)
plt.semilogy(x, np.abs((pr_result - reference)/reference), label="PR", alpha=0.5)

plt.legend()
plt.title(f"Relative error genlogistic ppf: $c={c}$")
plt.show()
```

</details>

*SF*

image

<details>

```
import numpy as np
from scipy import stats
from scipy import special as sc
import matplotlib.pyplot as plt

def genlogistic_logcdf(x, c):
    return -c * np.log1p(np.exp(-x))

def genlogistic_sf(x, c):
    return -sc.expm1(genlogistic_logcdf(x, c))

c = 0.02
x = np.linspace(0, 500, 100)
plt.semilogy(x, stats.genlogistic.sf(x, c), label="main", ls="dashed")
plt.semilogy(x, genlogistic_sf(x, c), label="PR", ls="dashdot")
plt.legend()
plt.title(f"Genlogistic survival function: $c={c}$")
plt.show()
```

</details>

image

<details>

```
import numpy as np
from scipy import stats
from scipy import special as sc
import matplotlib.pyplot as plt
from mpmath import mp

def genlogistic_logcdf(x, c):
    return -c * np.log1p(np.exp(-x))

def genlogistic_sf(x, c):
    return -sc.expm1(genlogistic_logcdf(x, c))

@np.vectorize
def sf_genlogistic_mpmath(x, c):
    x = mp.mpf(x)
    c = mp.mpf(c)
    return float(mp.one - (mp.one + mp.exp(-x))**(-c))

x = np.linspace(0, 500, 100)
c = 0.02
main_result = stats.genlogistic.sf(x, c)
custom_result = genlogistic_sf(x, c)
reference = np.asarray(sf_genlogistic_mpmath(x, c), np.float64)

plt.semilogy(x, np.abs((main_result - reference)/reference), label="main", alpha=0.5)
plt.semilogy(x, np.abs((custom_result - reference)/reference), label="PR", alpha=0.5)
plt.legend()
plt.title(f"Relative error genlogistic survival function: $c={c}$")
plt.show()
```

</details>

*ISF*

image

<details>
```
import numpy as np
from scipy import stats
from scipy import special as sc
import matplotlib.pyplot as plt
from mpmath import mp

mp.dps = 500

@np.vectorize
def genlogistic_isf_mpmath(x, c):
    x = mp.mpf(x)
    c = mp.mpf(c)
    return -mp.log((mp.one - x)**(-mp.one/c) - mp.one)

def genlogistic_isf(q, c):
    return -np.log(sc.powm1(1 - q, -1.0/c))

x = np.linspace(1e-30, 0.9999999999, 1000)
c = 20.
reference = np.asarray(genlogistic_isf_mpmath(x, c), np.float64)
main_result = stats.genlogistic.isf(x, c)
powm1_result = genlogistic_isf(x, c)


plt.semilogy(x, np.abs((main_result - reference)/reference), label="main", alpha=0.5)
plt.semilogy(x, np.abs((powm1_result - reference)/reference), label="PR", alpha=0.5)

plt.legend()
plt.title(f"Relative error genlogistic inv. survival function: $c={c}$")
plt.show()
```

</details>

### CODE REVIEW

**Review by mdhaber — CHANGES_REQUESTED** [review]
Looks accurate, but two suggestions.

Also, I think there might be an advantage to implementing `_logsf` using `logsumexp` while you're at it, but it's up to you.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -1435,13 +1435,53 @@ def test_logpdf(self, x, expected):
...
+    # mp.dps = 50
+    # def genlogistic_ppf_mpmath(x, c):
+    #     x = mp.mpf(x)
+    #     c = mp.mpf(c)
+    #     return -mp.log(x**(-mp.one/c) - mp.one)
+
+    @pytest.mark.parametrize('q, c, ref', [(0.1, 20, 2.10358298091805),
+                                           (0.9, 0.2, 0.3659913776736257)])
+    def test_ppf(self, q, c, ref):
+        assert_allclose(stats.genlogistic.ppf(q, c), ref, rtol=5e-15)
```
[the remark this answers is not in the document]
Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py]: To me, `logcdf` was always more of a helper to get the actual `cdf` or `sf`. Will add tests though, too, does not hurt.

**Comment by dschmitz89** [comment]
> Looks accurate, but two suggestions.
> 
> Also, I think there might be an advantage to implementing `_logsf` using `logsumexp` while you're at it, but it's up to you.

How would that `_logsf` look like?

**Comment by mdhaber** [comment]
Well, I thought it would go something like:
```
# Oops - this gets `logcdf` from `logsf`
from scipy import stats, special
x = [-8, 0, 8]
logsf = stats.norm().logsf(x)
log_one, logsf = np.broadcast_arrays(np.log(1), logsf)
logcdf = special.logsumexp([log_one + np.pi*1j, logsf ], axis=0)
print(np.real(logcdf))  # [-3.51036013e+01 -6.93147181e-01 -6.66133815e-16]
print(stats.norm().logcdf(x))  # [-3.50134372e+01 -6.93147181e-01 -6.22096057e-16]
print(np.log(stats.norm().cdf(x)))  # [-3.50134372e+01 -6.93147181e-01 -6.66133815e-16]
```
but it doesn't look it helps us any:

```
from scipy import stats, special
x = [-10, 0, 10]
logsf = stats.norm().logsf(x)
log_one, logsf = np.broadcast_arrays(np.log(1), logsf)
logcdf = special.logsumexp([log_one + np.pi*1j, logsf ], axis=0)
print(np.real(logcdf))  # [-3.66387090e+01 -6.93147181e-01  4.35396463e-79]
print(stats.norm().logcdf(x))  # [-5.32312852e+01 -6.93147181e-01 -7.61985302e-24]
print(np.log(stats.norm().cdf(x)))  # [-53.23128515  -0.69314718   0.        ]
```
so never mind.

**Comment by dschmitz89** [comment]
> Well, I thought it would go something like:
> 
> ```python
> # Oops - this gets `logcdf` from `logsf`
> from scipy import stats, special
> x = [-8, 0, 8]
> logsf = stats.norm().logsf(x)
> log_one, logsf = np.broadcast_arrays(np.log(1), logsf)
> logcdf = special.logsumexp([log_one + np.pi*1j, logsf ], axis=0)
> print(np.real(logcdf))  # [-3.51036013e+01 -6.93147181e-01 -6.66133815e-16]
> print(stats.norm().logcdf(x))  # [-3.50134372e+01 -6.93147181e-01 -6.22096057e-16]
> print(np.log(stats.norm().cdf(x)))  # [-3.50134372e+01 -6.93147181e-01 -6.66133815e-16]
> ```
> 
> but it doesn't look like it helps us any:
> 
> ```python
> from scipy import stats, special
> x = [-10, 0, 10]
> logsf = stats.norm().logsf(x)
> log_one, logsf = np.broadcast_arrays(np.log(1), logsf)
> logcdf = special.logsumexp([log_one + np.pi*1j, logsf ], axis=0)
> print(np.real(logcdf))  # [-3.66387090e+01 -6.93147181e-01  4.35396463e-79]
> print(stats.norm().logcdf(x))  # [-5.32312852e+01 -6.93147181e-01 -7.61985302e-24]
> print(np.log(stats.norm().cdf(x)))  # [-53.23128515  -0.69314718   0.        ]
> ```
> 
> so never mind.

That `special.logsumexp([log_one + np.pi*1j, logsf`])` is a good trick though. I am sure I can put it to good use one day, thanks for sharing.

**Comment by mdhaber** [comment]
It was essential in the recent implementation of `truncnorm`. `logsumexp` is supposed to take weights `b` that you can use to do subtraction, but they did not work well for me; directly adding `np.pi*1j` to the logarithm of the number to be subtracted was much more effective for me.

We don't need to do this in this PR, but here are some thoughts for getting an accurate `logsf` from `logcdf`.
When `logcdf` is very negative (i.e. `cdf` is close to zero), we run into trouble because `sf = 1-cdf` is indistinguishable from `1`, so `logsf` will be zero.

```python3
logcdf = -200
logsf = np.log(1 - np.exp(logcdf))  # 0.0
```
But we can do about as well as possible using `log1p`.
```
logsf = np.log1p(-np.exp(logcdf))  # -1.3838965267367376e-87
```
Typically, exponentiating a very negative `logcdf` is bad because it can underflow. But if `np.exp(logcdf)` underflows, then the true value of `logcdf` would underflow, anyway, so we're not losing anything.

When `logcdf` is very close to zero (i.e. `cdf` is close to one), `sf = 1 - cdf` can't get very close to zero without being exactly zero because `cdf` can only represent numbers ~`1e-16` from 1, and `logsf` becomes `-inf`.
```python3
logcdf = -1e-200
logsf = np.log(1 - np.exp(logcdf))  # -inf
```
But we can do about as well as possible using `expm1`:
```
# `+0j` avoids the NaNs you get with negative numbers
logsf = np.real(np.log(-special.expm1(logcdf+0j)))  # -460.51701859880916
```
This won't go to `-inf` unless `logcdf` underflows.

```python3
import numpy as np
from scipy import stats, special
from scipy._lib._util import _lazywhere
from mpmath import mp
mp.dps = 300

x = [-20, -10, -5, 0, 5, 10, 20]

def logsf(dist, x):
    logcdf = dist.logcdf(x)
    
    def f1(x):
        return np.log1p(-np.exp(x))
    
    def f2(x):
        return np.real(np.log(-special.expm1(x+0j)))
    
    return _lazywhere(logcdf < -0.01, (logcdf, ), f=f1, f2=f2)


@np.vectorize
def logsf_ref(x):
    x = mp.mpf(float(x))
    return float(mp.log(1 - mp.ncdf(x)))

res = logsf(stats.norm(), x)
ref = logsf_ref(x)
np.testing.assert_allclose(res, ref)
```

**Review by mdhaber — APPROVED** [review]
Again, just condensed the tests. Now, the reference CDF implementation was written from the definition of CDF that predates the PR, showing that all these new methods are consistent with that and accurate.

**Pull request merged**